test/cpp/naming/resolver_component_tests_runner_invoker_common.cc

```diff
@@ -0,0 +1,137 @@
+/*
+ *
+ * Copyright 2017 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <grpc/grpc.h>
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/string_util.h>
+#include <signal.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <gflags/gflags.h>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include "test/cpp/util/subprocess.h"
+#include "test/cpp/util/test_config.h"
+
+#include "test/cpp/naming/resolver_component_tests_runner_invoker_common.h"
+
+extern "C" {
+#include "test/core/util/port.h"
+}
+
+using grpc::SubProcess;
+
+static volatile sig_atomic_t abort_wait_for_child = 0;
+
+static void sighandler(int sig) { abort_wait_for_child = 1; }
+
+static void register_sighandler() {
+  struct sigaction act;
+  memset(&act, 0, sizeof(act));
+  act.sa_handler = sighandler;
+  sigaction(SIGINT, &act, NULL);
+  sigaction(SIGTERM, &act, NULL);
+}
+
+namespace {
+
+const int kTestTimeoutSeconds = 60 * 2;
+
+void RunSigHandlingThread(SubProcess *test_driver, gpr_mu *test_driver_mu,
+                          gpr_cv *test_driver_cv, int *test_driver_done) {
+  for (size_t i = 0; i < kTestTimeoutSeconds && !abort_wait_for_child; i++) {
```

Comment:
This seems a little inexact, since the time spent processing the body of the loop will be non-zero, so the total timeout time will be more than `kTestTimeoutSeconds`.  Instead, how about computing the overall deadline before starting the loop and then setting the timeout passed to `gpr_cv_wait()` to be min(1 second, time remaining before deadline)?

Reply:
Just checking, I am a little hesitant since that would cause a response to `SIGINT`/`SIGTERM` to be delayed until the test timeout. But I would be ok, do we want to go ahead with that?

Reply:
I don't think this would cause the response to `SIGINIT` / `SIGTERM` to be delayed.  The overall mechanics of the loop would not be different -- we would still be blocking for no more than 1 second at a time.

What I'm thinking is something like this:

```
gpr_timespec deadline = gpr_time_add(
    gpr_now(GPR_CLOCK_MONOTONIC),
    gpr_time_from_seconds(kTestTimeoutSeconds, GPR_TIMESPAN));
while (true) {
  gpr_timespec now = gpr_now(GPR_CLOCK_MONOTONIC);
  if (gpr_time_cmp(now, deadline) <= 0) break;
  gpr_mu_lock(test_driver_mu);
  if (*test_driver_done) {
    gpr_mu_unlock(test_driver_mu);
    return;
  }
  gpr_timespan wait_deadline = gpr_time_min(
      deadline,
      gpr_time_add(now, gpr_time_from_seconds(1, GPR_TIMESPAN)));
  gpr_cv_wait(test_driver_cv, test_driver_mu, wait_deadline);
  gpr_mu_unlock(test_driver_mu);    
}
```

Reply:
I see, misunderstood the earlier suggestion and didn't realize it could still poll every second. thanks, done